understanding/20/contrast-enhanced.html

```diff
@@ -73,13 +73,19 @@ <h2>Intent of Contrast (Enhanced)</h2>
       </div>
 
       <div class="note">
-         <p>Because authors do not have control over user settings for font smoothing/anti-aliasing, when evaluating this
-            Success Criterion, refer to the foreground and background colors obtained from the user agent, or the underlying
-            markup and stylesheets, rather than the text as presented on screen.</p>
-         <p>Due to anti-aliasing, particularly thin or unusual fonts may be rendered by user agents with a much fainter
-            color than the actual text color defined in the underlying CSS. This can lead to situations where text has
-            a contrast ratio that nominally passes the Success Criterion, but has a much lower contrast in practice.
-            In these cases, best practice would be for authors to choose a font with stronger/thicker lines,
+         <p>User agents which render text will often apply anti-aliasing, font smoothing, and other techniques
+            which can cause some pixels of text to render in fainter colors than those specified by authors.
+            When evaluating this Success Criterion, this is usually not relevant; as long as the applied font,
+            size, and weight ensure that some pixel of each character reaches the author-specified color, evaluators
```

Comment:
It should be carefully considered what exactly is required here: is just one pixel with sufficient contrast in a letter really enough? Imagine the letter x. Due to the slanted lines, the contrasts in the entire letter are often not sufficient, except at the intersection of the two lines. So I have a dot in the middle, but it doesn't help me to recognize the letter itself. If a sensible contrast requirement is to be established here, I would demand that every relevant line of a letter has at least one dot with sufficient contrast at every point, i.e. in the case of the x, for example, both lines over their entire length, except for the serifs.

Reply:
I don't know how we can be clear about that. There is no defined method for deciding which pixels to compare to which other pixels when measuring contrast, and then how many and by how much those are allowed to follow below the minimum before the text is considered to be failing.

Reply:
If I recall correctly - the ACT rule is where the one pixel check came from and that rule clarifies that if the check passes further testing may be needed to determine conformance with WCAG SC 1.4.3 which does not specify.  When the AG group talked about this many years ago it was said that all pixels needed to contrast.   I had thrown out a suggestion previously when we were discussing the ACT rule that we might follow the similar principle in 1.4.11 where we require the icon;/letter be identifiable with only contrasting pixels.  That is - remove the non-contrasting pixels and can you still tell what letter it is.  If not - it would not pass- if you can then it would pass.

Reply:
One example: I have chosen the font Arial and font size 16 px (this is neither a thin nor a small font). Colours are black (#000) on grey (#757575). Contrast is 4.6:1. 

I have chosen the letters x, y and z. In the two illustrations, the representation is greatly enlarged. In the second illustration, the pixels with sufficient contrast are marked in yellow:
* for x: 3 pixels in the centre and one pixel at the bottom right
* for the y: no pixel
* for z: 5 pixels in the upper and 6 in the lower bar, but no pixel in the horizontal bar. 

The letters cannot be identified using the pixels with sufficient contrast. 

There is also the problem that the pixels with sufficient contrast cannot be recognised with the eye, i.e. if I want to measure the actual contrasts, I have to measure each pixel of each letter individually. 

![xyz](https://github.com/w3c/wcag/assets/52748763/1af3d627-4fd3-4003-9b59-d7d4c78c1ca4)
![xyz_contrast](https://github.com/w3c/wcag/assets/52748763/163f8652-2ad4-4453-b1ee-56fae1b00cbf)

Reply:
And I think every pixel in every letter really needs to be measured, because even with the same letters the contrasts are different depending on their position in the word. The following two images show the letters "by y". In the second illustration, the pixels with sufficient contrast are marked in yellow:
* for b: 2 pixels at the top and 2 pixels at the bottom, 2 pixels in the horizontal bar (so it looks like the letter c)
* for the first y: 3 pixels at the bottom
* for the second y: no pixel

![byy](https://github.com/w3c/wcag/assets/52748763/31735890-59b7-48e6-9020-51c106e06c5f)
![byy_contrast](https://github.com/w3c/wcag/assets/52748763/43957358-5868-4551-ba70-f734999e05d8)

Reply:
I would like to agree with the opinion of @mraccess77: if we measure the pixels, then it only makes sense if all pixels that have sufficient contrast are sufficient to recognise the letter or graphic object. The problem is that this can lead to different results depending on the anti-aliasing settings, which is why the WCAG uses the CSS values and not the actual pixels to calculate the contrasts. For content that is not based on CSS (e.g. pixel graphics), there is no other option than to measure the pixels. But even with pixel graphics, the question would be how I measure: The displayed graphic on the page (which can be enlarged or reduced in size) or the original graphic?

Reply:
These images are a really interesting case to discuss because they are demonstrations of cases where there is a common, practical, meaningful difference between the **perceived color as rendered on hardware** and **the color of individual pixels as measured by a color picker**. These cases where the left side of a glyph is tinted red and the right side is tinted blue are demonstrating a technique called [Subpixel rendering](https://en.wikipedia.org/wiki/Subpixel_rendering), whereby the text rendering engine is taking advantage of knowing the specific type of screen displaying the text to essentially render pixels of the text "in between" pixels of the screenshot. The user-perceived color in a case like this is going to be closer to the specified color than what a tester could evaluate using a pixel-wise color picker.

Notably, this is *not* the same as anti-aliasing; the note in the contrast ratio definition that says you can ignore anti-aliasing wouldn't apply to this technique.

I think this is a very strong argument against this PR as-written and in favor of allowing testing against specified colors *even if an assessment based on color picking would result in a failure*.

Reply:
> the note in the contrast ratio definition that says you can ignore anti-aliasing wouldn't specifically apply to subpixel rendering.

when that note was written in 2000ish or so, there was no subpixel rendering, i'd hazard to say...

Reply:
Thanks @JAWS-test for those examples!